### PR TITLE
Use improved criticality estimates for the first routing iteration

### PR DESCRIPTION
**Pull request by kmurray**
#### Description

Traditionally VPR has routed all connections during the 1st router iteration as timing critical (i.e. criticality = 1). While this helps ensure good initial delay estimates it has a CPU-time cost.

In particular, since the spatial-lookup for High Fanout (HF) nets is only active for less timing critical nets (i.e. criticality < 0.9) using high criticality can have a significant run-time cost on designs with large high-fanout nets.

Currently this PR only implements this functionality, which is off by default.

#### Related Issue
#1073 

#### Motivation and Context

If we can use more realistic criticality estimates we can ensure that less-timing critical high-fanout  nets use the more efficient spatial-lookup based approach.

Also, for designs which aren't heavily congested the 1st routing iteration is usually the longest, so reducing it's run-time is desirable.

#### How Has This Been Tested?

Currently tested locally on vtr_reg_basic, vtr_reg_strong. Also on the Titan benchmarks with `--router_lookahead map --router_initial_timing lookahead --router_update_lower_bound_delays on`, showing ~6% route time reduction on average, with larger reductions on large designs with high-fanout nets (e.g. 35% route-time reduction on gaussianblur).

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by kmurray** [comment]
One item for consideration is what the new defaults should be for these settings.

Some options include:
* `--router_initial_timing lookahead --router_update_lower_bound_delays on` always
* `--router_initial_timing lookahead --router_update_lower_bound_delays on` when the map lookahead is used (not sure how the classic lookahead will behave if asked to estimate delays from a SOURCE)

**Comment by kmurray** [comment]
@litghost Since you are routing (presumably high-fanout) constant nets this may be of interest, as during the 1st routing iteration those would currently be routed using the 'slow' (non-spatial look-up) approach. 

Additionally, using more reasonable criticalities for the 1st iteration keeps wirelength much better in check (only estimated timing critical connections use a star-like routing).

**Comment by kmurray** [comment]
QoR on Titan: [comparison_output.xlsx](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/4154702/comparison_output.xlsx)

Largest gains are on the largest Titan design `guassianblur` where the 1st iteration route time is cut from 1461 sec to 188 sec:

*Baseline*
```# Routing
RR graph channel widths unchanged, skipping RR graph rebuild
Confirming router algorithm: TIMING_DRIVEN.
## Initializing router criticalities
Initial Net Connection Criticality Histogram:
[        0:      0.1)       0 (  0.0%) |
[      0.1:      0.2)       0 (  0.0%) |
[      0.2:      0.3)       0 (  0.0%) |
[      0.3:      0.4)       0 (  0.0%) |
[      0.4:      0.5)       0 (  0.0%) |
[      0.5:      0.6)       0 (  0.0%) |
[      0.6:      0.7)       0 (  0.0%) |
[      0.7:      0.8)       0 (  0.0%) |
[      0.8:      0.9)       0 (  0.0%) |
[      0.9:        1) 4171299 (100.0%) |*******************************************
## Initializing router criticalities took 0.25 seconds (max_rss 24120.6 MiB, delta_rss +0.0 MiB)
---- ------ ------- ---- ------- ------- ------- ----------------- --------------- -------- ---------- ---------- ---------- ---------- --------
Iter   Time    pres  BBs    Heap  Re-Rtd  Re-Rtd Overused RR Nodes      Wirelength      CPD       sTNS       sWNS       hTNS       hWNS Est Succ
      (sec)     fac Updt    push    Nets   Conns                                       (ns)       (ns)       (ns)       (ns)       (ns)     Iter
---- ------ ------- ---- ------- ------- ------- ----------------- --------------- -------- ---------- ---------- ---------- ---------- --------
   1 1461.3     0.0    0 9.3e+09 1211052 4070606 2605720 ( 5.321%) 3.8e+07 (54.2%)  870.186 -1.005e+08   -869.186      0.000      0.000      N/A
   2  210.7     0.5 6491 1.8e+09 1039132 3883136  594297 ( 1.214%) 3.0e+07 (42.9%)  871.860 -1.295e+08   -870.860      0.000      0.000      N/A
   3  156.5     0.6 1670 1.2e+09  448545 2897051  466922 ( 0.954%) 3.0e+07 (42.8%)  871.989 -1.269e+08   -870.989      0.000      0.000      N/A
   4  167.8     0.8 1254 1.1e+09  382960 2567524  395814 ( 0.808%) 3.0e+07 (43.0%)  871.487 -1.295e+08   -870.487      0.000      0.000      N/A
   5  183.5     1.1 1347 1.1e+09  318555 2320503  281206 ( 0.574%) 3.0e+07 (43.2%)  871.520 -1.280e+08   -870.520      0.000      0.000      N/A
   6  203.2     1.4 1571 1.0e+09  232826 1852527  198711 ( 0.406%) 3.1e+07 (43.6%)  871.540 -1.286e+08   -870.540      0.000      0.000      N/A
   7  175.3     1.9 1716 8.8e+08  168188 1319144  132691 ( 0.271%) 3.1e+07 (43.9%)  871.629 -1.301e+08   -870.629      0.000      0.000      N/A
   8  154.8     2.4 1595 7.0e+08  112151  912815   79600 ( 0.163%) 3.1e+07 (44.2%)  872.304 -1.306e+08   -871.304      0.000      0.000      N/A
   9  123.6     3.1 1671 5.0e+08   72860  609851   48159 ( 0.098%) 3.1e+07 (44.5%)  872.174 -1.299e+08   -871.174      0.000      0.000      N/A
  10   94.4     4.1 1430 3.4e+08   44033  377372   26908 ( 0.055%) 3.1e+07 (44.7%)  872.209 -1.305e+08   -871.209      0.000      0.000       33
  11   64.0     5.3 1092 1.9e+08   23219  208316   11593 ( 0.024%) 3.1e+07 (44.8%)  872.193 -1.304e+08   -871.193      0.000      0.000       30
  12   46.8     6.9  715 1.1e+08   12684  111705    5337 ( 0.011%) 3.2e+07 (44.9%)  872.075 -1.307e+08   -871.075      0.000      0.000       28
  13   35.5     9.0  375 5.3e+07    7271   49725    1990 ( 0.004%) 3.2e+07 (44.9%)  872.106 -1.307e+08   -871.106      0.000      0.000       26
  14   29.7    11.6  161 2.8e+07    5102   27361     963 ( 0.002%) 3.2e+07 (45.0%)  872.211 -1.308e+08   -871.211      0.000      0.000       24
  15   26.7    15.1   73 1.7e+07    4172   16941     394 ( 0.001%) 3.2e+07 (45.0%)  872.202 -1.308e+08   -871.202      0.000      0.000       23
  16   24.4    19.7   26 1.2e+07    3767   14284     205 ( 0.000%) 3.2e+07 (45.0%)  872.218 -1.308e+08   -871.218      0.000      0.000       23
  17   23.6    25.6   10 1.1e+07    3634   13096     123 ( 0.000%) 3.2e+07 (45.0%)  872.216 -1.308e+08   -871.216      0.000      0.000       23
  18   23.1    33.3    3 1.1e+07    3581   12875      74 ( 0.000%) 3.2e+07 (45.0%)  872.215 -1.308e+08   -871.215      0.000      0.000       23
  19   23.1    43.3    0 1.0e+07    3543   12740      42 ( 0.000%) 3.2e+07 (45.0%)  872.225 -1.308e+08   -871.225      0.000      0.000       23
  20   23.0    56.2    0 1.0e+07    3528   12702      33 ( 0.000%) 3.2e+07 (45.0%)  872.225 -1.308e+08   -871.225      0.000      0.000       24
  21   24.8    73.1    0 9895488    3517   12681      23 ( 0.000%) 3.2e+07 (45.0%)  872.225 -1.308e+08   -871.225      0.000      0.000       25
  22   29.4    95.0    0 9693476    3510   12652      11 ( 0.000%) 3.2e+07 (45.0%)  872.225 -1.308e+08   -871.225      0.000      0.000       25
  23   23.4   123.5    0 9497695    3509   12650       3 ( 0.000%) 3.2e+07 (45.0%)  872.225 -1.308e+08   -871.225      0.000      0.000       26
  24   23.1   160.6    0 9282733    3500   12633       0 ( 0.000%) 3.2e+07 (45.0%)  872.225 -1.308e+08   -871.225      0.000      0.000       25
Restoring best routing
Critical path: 872.225 ns
Successfully routed after 24 routing iterations.
Final Net Connection Criticality Histogram:
[        0:      0.1) 3380746 ( 81.0%) |*******************************************
[      0.1:      0.2)    7313 (  0.2%) |
[      0.2:      0.3)    1415 (  0.0%) |
[      0.3:      0.4)  161442 (  3.9%) |**
[      0.4:      0.5)  215142 (  5.2%) |***
[      0.5:      0.6)  202034 (  4.8%) |***
[      0.6:      0.7)  130347 (  3.1%) |**
[      0.7:      0.8)   18792 (  0.5%) |
[      0.8:      0.9)    3773 (  0.1%) |
[      0.9:        1)   50295 (  1.2%) |*
Router Stats: total_nets_routed: 4114839 total_connections_routed: 21340890 total_heap_pushes: 18503812737 total_heap_pops: 1706382106
# Routing took 3401.38 seconds (max_rss 26096.9 MiB, delta_rss +3850.7 MiB)
```

*This PR with --router_initial_timing lookahead --router_update_lower_bound_delays on*
```
# Routing
RR graph channel widths unchanged, skipping RR graph rebuild
Confirming router algorithm: TIMING_DRIVEN.
## Initializing router criticalities
Warning 924: Failed to get delay estimates for initial criticality for 0 net connections (assumed zero delay)
Initial Net Connection Criticality Histogram:
[        0:      0.1) 3387687 ( 81.2%) |*******************************************
[      0.1:      0.2)  400441 (  9.6%) |*****
[      0.2:      0.3)  328504 (  7.9%) |****
[      0.3:      0.4)      55 (  0.0%) |
[      0.4:      0.5)     137 (  0.0%) |
[      0.5:      0.6)     229 (  0.0%) |
[      0.6:      0.7)     131 (  0.0%) |
[      0.7:      0.8)     125 (  0.0%) |
[      0.8:      0.9)    3485 (  0.1%) |
[      0.9:        1)   50505 (  1.2%) |*
## Initializing router criticalities took 58.78 seconds (max_rss 24281.4 MiB, delta_rss +161.6 MiB)
---- ------ ------- ---- ------- ------- ------- ----------------- --------------- -------- ---------- ---------- ---------- ---------- --------
Iter   Time    pres  BBs    Heap  Re-Rtd  Re-Rtd Overused RR Nodes      Wirelength      CPD       sTNS       sWNS       hTNS       hWNS Est Succ
      (sec)     fac Updt    push    Nets   Conns                                       (ns)       (ns)       (ns)       (ns)       (ns)     Iter
---- ------ ------- ---- ------- ------- ------- ----------------- --------------- -------- ---------- ---------- ---------- ---------- --------
   1  188.7     0.0    0 1.9e+09 1211052 4070606 2372512 ( 4.845%) 2.9e+07 (41.9%)  870.155 -1.150e+08   -869.155      0.000      0.000      N/A
   2  216.9     0.5 3253 1.9e+09 1126870 3972464  604547 ( 1.235%) 3.0e+07 (43.0%)  871.844 -1.221e+08   -870.844      0.000      0.000      N/A
   3  162.3     0.6 1678 1.2e+09  440344 2792989  462329 ( 0.944%) 3.0e+07 (42.8%)  872.093 -1.253e+08   -871.093      0.000      0.000      N/A
   4  170.2     0.8 1221 1.1e+09  374390 2464236  388732 ( 0.794%) 3.0e+07 (42.9%)  872.246 -1.266e+08   -871.246      0.000      0.000      N/A
   5  189.1     1.1 1327 1.1e+09  310058 2198304  276160 ( 0.564%) 3.0e+07 (43.2%)  871.473 -1.264e+08   -870.473      0.000      0.000      N/A
   6  188.4     1.4 1464 1.0e+09  226011 1756643  194376 ( 0.397%) 3.1e+07 (43.6%)  871.905 -1.272e+08   -870.905      0.000      0.000      N/A
   7  176.3     1.9 1704 8.9e+08  165582 1287515  130760 ( 0.267%) 3.1e+07 (43.9%)  871.911 -1.270e+08   -870.911      0.000      0.000      N/A
   8  159.6     2.4 1648 7.1e+08  112224  899829   79301 ( 0.162%) 3.1e+07 (44.2%)  872.100 -1.276e+08   -871.100      0.000      0.000      N/A
   9  127.2     3.1 1667 5.0e+08   71493  582480   47142 ( 0.096%) 3.1e+07 (44.4%)  871.837 -1.274e+08   -870.837      0.000      0.000      N/A
  10   99.1     4.1 1428 3.5e+08   43334  371449   26989 ( 0.055%) 3.1e+07 (44.6%)  871.776 -1.277e+08   -870.776      0.000      0.000       33
  11   74.0     5.3 1130 2.1e+08   23900  217919   11964 ( 0.024%) 3.1e+07 (44.8%)  872.068 -1.274e+08   -871.068      0.000      0.000       31
  12   48.4     6.9  720 1.0e+08   11751   93994    4686 ( 0.010%) 3.1e+07 (44.8%)  872.217 -1.278e+08   -871.217      0.000      0.000       28
  13   38.0     9.0  371 5.3e+07    6975   47833    1914 ( 0.004%) 3.2e+07 (44.9%)  872.216 -1.279e+08   -871.216      0.000      0.000       25
  14   31.7    11.6  160 2.7e+07    4932   25347     818 ( 0.002%) 3.2e+07 (44.9%)  872.202 -1.279e+08   -871.202      0.000      0.000       24
  15   27.3    15.1   54 1.7e+07    4054   16222     382 ( 0.001%) 3.2e+07 (44.9%)  872.212 -1.279e+08   -871.212      0.000      0.000       23
  16   24.7    19.7   20 1.2e+07    3733   14393     209 ( 0.000%) 3.2e+07 (44.9%)  872.228 -1.279e+08   -871.228      0.000      0.000       23
  17   23.8    25.6    8 1.1e+07    3615   13195     124 ( 0.000%) 3.2e+07 (44.9%)  872.228 -1.279e+08   -871.228      0.000      0.000       22
  18   23.5    33.3    2 1.1e+07    3560   12890      73 ( 0.000%) 3.2e+07 (44.9%)  872.232 -1.279e+08   -871.232      0.000      0.000       23
  19   23.6    43.3    1 1.0e+07    3544   12737      44 ( 0.000%) 3.2e+07 (44.9%)  872.501 -1.279e+08   -871.501      0.000      0.000       23
  20   23.7    56.2    0 9981441    3523   12674      27 ( 0.000%) 3.2e+07 (44.9%)  872.501 -1.279e+08   -871.501      0.000      0.000       24
  21   23.4    73.1    1 9750449    3509   12649      16 ( 0.000%) 3.2e+07 (44.9%)  872.605 -1.279e+08   -871.605      0.000      0.000       24
  22   23.4    95.0    0 9566611    3504   12646       2 ( 0.000%) 3.2e+07 (44.9%)  872.605 -1.279e+08   -871.605      0.000      0.000       25
  23   23.6   123.5    0 9355259    3498   12622       0 ( 0.000%) 3.2e+07 (44.9%)  872.605 -1.279e+08   -871.605      0.000      0.000       24
Restoring best routing
Critical path: 872.605 ns
Successfully routed after 23 routing iterations.
Final Net Connection Criticality Histogram:
[        0:      0.1) 3380749 ( 81.0%) |*******************************************
[      0.1:      0.2)    7405 (  0.2%) |
[      0.2:      0.3)    1032 (  0.0%) |
[      0.3:      0.4)  162461 (  3.9%) |**
[      0.4:      0.5)  224038 (  5.4%) |***
[      0.5:      0.6)  199550 (  4.8%) |***
[      0.6:      0.7)  127715 (  3.1%) |**
[      0.7:      0.8)   14364 (  0.3%) |
[      0.8:      0.9)    3696 (  0.1%) |
[      0.9:        1)   50289 (  1.2%) |*
Router Stats: total_nets_routed: 4161456 total_connections_routed: 20901636 total_heap_pushes: 11151762511 total_heap_pops: 1506002944
# Routing took 2197.65 seconds (max_rss 26065.7 MiB, delta_rss +3819.9 MiB)
```

File: vpr/src/route/router_lookahead_map.cpp [hunk 1]
```diff
@@ -215,37 +243,76 @@ float get_lookahead_map_cost(int from_node_ind, int to_node_ind, float criticali
...
 
-    /* now get the expected cost from our lookahead map */
-    int from_chan_index = 0;
-    if (from_type == CHANY) {
-        from_chan_index = 1;
-    }
+    float expected_cost = std::numeric_limits<float>::infinity();
+
+    e_rr_type from_type = from_node.type();
+    if (from_type == SOURCE || from_type == OPIN) {
```
Comment (litghost) [inline vpr/src/route/router_lookahead_map.cpp]: Can you explain what is going on with this change?  It seems orthogonal to the overall PR?

**Comment by litghost** [comment]
This PR modifies the data storage for the map lookahead.  Can you please first merge #1100, then update the serialization to capture the new data?

File: vpr/src/route/router_lookahead_map.cpp (hunk 1 above)
Comment (kmurray) [reply vpr/src/route/router_lookahead_map.cpp]: In order to get reasonable delay estimates before the 1st routing iteration (to figure out criticalities) we use the map lookahead. However the map lookahead (before these changes) would crash if you tried to estimate delay from something other than a CHANX/CHANY.

This extends the map lookahead to behave reasonably for estimating delays from SOURCEs (which is used for pre-routing delay estimates). See comments with e2551a5 and 04f7faf.

Comment (litghost) [reply vpr/src/route/router_lookahead_map.cpp]: Comments in the commits are useful, but what you just wrote really ought to be in the code.

**Comment by kmurray** [comment]
> This PR modifies the data storage for the map lookahead. Can you please first merge #1100, then update the serialization to capture the new data?

The actual map lookahead from wires is unchanged so that serialization will be the same. The side look-up table from pins to wires is small and very cheap to compute so I'd suggest just recomputing rather than serializing it.

**Comment by litghost** [comment]
> The actual map lookahead from wires is unchanged so that serialization will be the same. The side look-up table from pins to wires is small and very cheap to compute so I'd suggest just recomputing rather than serializing it.

Would you please measure how long it takes to compute it?  If it takes less than 1 second, I agree, we can just compute it live.

File: vpr/src/route/router_lookahead_map.cpp (hunk 1 above)
Comment (kmurray) [reply vpr/src/route/router_lookahead_map.cpp]: Agreed I'll put some more comments in.

**Comment by kmurray** [comment]
> Would you please measure how long it takes to compute it? If it takes less than 1 second, I agree, we can just compute it live

Yep, this PR also puts in timers to give the breakdown.

From the gaussianblur run:
```
# Computing router lookahead map
## Computing wire lookahead
## Computing wire lookahead took 1366.54 seconds (max_rss 22245.8 MiB, delta_rss +148.8 MiB)
## Computing src/opin lookahead
## Computing src/opin lookahead took 0.00 seconds (max_rss 22245.8 MiB, delta_rss +0.0 MiB)
# Computing router lookahead map took 1366.54 seconds (max_rss 22245.8 MiB, delta_rss +148.8 MiB)
```
which shows the src/opin lookahead takes < 0.01 seconds to compute.

**Comment by litghost** [comment]
> which shows the src/opin lookahead takes < 0.01 seconds to compute.

Awesome, nothing to worry about then.

File: vpr/src/route/router_lookahead.cpp [hunk 2]
```diff
@@ -92,7 +92,7 @@ float MapLookahead::get_expected_cost(int current_node, int target_node, const t
 
     t_rr_type rr_type = device_ctx.rr_nodes[current_node].type();
 
-    if (rr_type == CHANX || rr_type == CHANY) {
+    if (rr_type == CHANX || rr_type == CHANY || rr_type == SOURCE || rr_type == OPIN) {
```
Comment (litghost) [inline vpr/src/route/router_lookahead.cpp]: Can I suggest that the MapLookahead class definition should move to `router_lookahead_map.h/.cpp`?  It's been on my list to do, and this might be a good time.

Comment (kmurray) [reply vpr/src/route/router_lookahead.cpp]: Yep makes sense!
